### ISSUE TITLE
User's password is stored as plaintext in our database

### ISSUE THREAD

**Issue by xuhang57** [CCI-MOC/ui#62]
Need to hash the password before we store it

**Comment by zenhack** [comment]
The cluster passwords actually *can't* be hashed, because they're being used by the UI to authenticate against openstack. They thus *need* to be in a form that is usable as authentication (and thus equivalent to plaintext). We *are* hashing the password used to log into the UI.

Eventually we'll want to avoid storing the user's openstack password at all, but I think the plan there is federated keystone, which will take time.

**Comment by pns005** [comment]
So why can't the UI "re-hash" the stored "hashed" password before using it
to authenticate ?

This Icehouse wiki
<https://wiki.openstack.org/wiki/Security/Icehouse/Keystone>explains how
passwords are stored "hashed" ...

Piyanai

**Comment by xuhang57** [comment]
I agree with Ian that eventually, we should avoid storing user's password at all. All we need to do is to use their credentials to authenticate keystone/nova/neutron/swift/cinder/glance client service and we don't need those info other than that. 

Piyanai's wiki page is also right. However, I don't know if that's what we want so far.

**Comment by zenhack** [comment]
Quoting pns005 (2015-05-18 22:11:03)

>    So why can't the UI "re-hash" the stored "hashed" password before using
>    it to authenticate ?

The whole point of hashing passwords is to prevent an attacker who can 
read the password database from making use of the passwords. Storing the 
password hash gives you enough information to validate that a provided 
password is correct, but it *doesn't*  give you something you can 
actually log in with yourself. Again, that's the whole point. If we want 
the UI to be able to log in to keystone, we need to have the password in 
a form that's usable for that, which the hashed password isn't.

As I said, we are hashing the password the user uses to log in to the 
MOC UI, since we just need to check that for correctness, not log in 
with it ourselves. Based on the table you linked to (very last row), 
we're doing this exactly the same way keystone is.

-Ian

**Comment by pns005** [comment]
>Based on the table you linked to (very last row),
>we're doing this exactly the same way keystone is.

Are you talking about "sha512" ? So we are doing that ? If not, I don't
understand what you are talking about.
A snapshot would help.

>Storing the password hash gives you enough information to validate that a
provided
>password is correct, but it *doesn't* give you something you can
>actually log in with yourself.

Why not ? It might not be convenient to get both user name and password in
the same persistent area (mysql), but this is definitely a hole here, even
though our first phase of production deployment is for "trusted" users.

Regardless, having a non-hashed password in a persistent area/database is
just bad regardless of what the reason is.
We have a long list of security stuff that we have been talking about
revisiting. I don't remember if this is one of the topics or not.
If it's not, it should.

Btw, I don't claim to be an expert around security  - far from it. But I
know that at least one other OpenStack public cloud I am familiar with
(Grizzly release then) had user passwords in mysql but "hashed" first.

I can see that we will get into rat hole easily here via mailing list. And
we deinfitely have other higher priority items to work on this week.
So, I will put this on agenda of something to discuss in sprint 7.

Piyanai

**Comment by zenhack** [comment]
I'll try my hand at a clear explanation of the situation, but if there's 
still confusion, then yeah, we should just work it out in person.  
Apologies in advance if much of this is review, I just want to make sure 
we're all on the same page about what it is we're talking about.

What is meant by hashing a password is that we store a cryptographic
hash of the password, rather than the password itself. (We also add a
salt, but I'll not get into that; it's part of the same general best
practice). The idea behind this is that if someone gets their hands on
the database, they still can't actually log into the system, because the
hashes aren't reversible. In both the case of keystone, and the UI
(where we're able to), sha512 is the hash function used.

This scheme is designed for cases the password only needs to be used by
a service who is authenticating users, but doesn't need to actually be
able to log in *as* a given user. This is the common case; when you log 
in to Github, Github needs to check that your password is correct, but 
it doesn't need to have enough information to actually log in in order 
to do that. Instead, to check that your password is correct, it hashes 
the password you enter, and verifies that it comes out to the same value 
that's in the database.

The hash is irreversible; If it's working correctly, you can't recover 
the password from it.

There are two kinds of passwords that are used by the MOC UI:

* The password that a user uses to log in to the MOC UI. This one is 
  stored hashed and salted, according to standard best practices.
* The user's Openstack password, i.e. the password that must be provided 
  to keystone in order to get a token. The UI logs in on behalf of the 
  user, so it must have this in a form which is usable for that. Since 
  the whole point of hashing passwords is for them to *not* be usable 
  for this, hashing the password is a no-go.

Keystone stores the second password hashed and salted, in exactly the 
same way we stored the first. It's able to do this where we are not, 
since it never needs to actually provide the password, only verify it.

I think we're all uncomfortable with storing passwords in plaintext, but 
hashing the Openstack password and storing it isn't a viable solution.  
The long term goal is to have federated keystone, and not need to store 
any passwords in the UI's database at all, but we need federated 
keystone for that. I'm open to other options, but I don't recall us 
coming up with anything compelling the last time we talked about it.

-Ian
